runtime/src/main/java/org/corfudb/runtime/view/stream/QuorumStreamView.java

```diff
@@ -0,0 +1,116 @@
+       }
+       return res;
+    }
+
+
+    /**
+     * {@inheritDoc}
+     *
+     * In quorum stream view it is possible the offset to change, so we are returning the last token
+     * from the sequencer for this context id.
+     */
+    @Override
+    public long append(Object object,
+                       Function<TokenResponse, Boolean> acquisitionCallback,
+                       Function<TokenResponse, Boolean> deacquisitionCallback) {
+        // First, we get a token from the sequencer.
+        TokenResponse tokenResponse = runtime.getSequencerView()
+                .nextToken(Collections.singleton(ID), 1);
+        // We loop forever until we are interrupted, since we may have to
+        // acquire an address several times until we are successful.
+
+        boolean failover = runtime.getSequencerView().isFailOverSequencer();
+        int rank = 0;
+        while (true) {
+            // Next, we call the acquisitionCallback, if present, informing
+            // the client of the token that we acquired.
+            if (acquisitionCallback != null) {
+                if (!acquisitionCallback.apply(tokenResponse)) {
+                    // The client did not like our token, so we end here.
+                    // We'll leave the hole to be filled by the client or
+                    // someone else.
+                    log.debug("Acquisition rejected token={}", tokenResponse);
+                    return -1L;
+                }
+            }
+
+            // >0 means a recovery mode
+            recoveryModeRankThreadLocal.set(rank);
+
+            // Now, we do the actual write. We could get an overwrite
+            // exception here - any other exception we should pass up
+            // to the client.
+            try {
+                runtime.getAddressSpaceView()
+                        .write(tokenResponse.getToken(),
+                                Collections.singleton(ID),
+                                object,
+                                tokenResponse.getBackpointerMap(),
+                                tokenResponse.getStreamAddresses());
+
+                // The write completed successfully, so we return this
+                // address to the client.
+                return tokenResponse.getToken();
+            } catch (OverwriteException oe) {
+                log.trace("Overwrite occurred at {}", tokenResponse);
+                // We got overwritten, so we call the deacquisition callback
+                // to inform the client we didn't get the address.
+                if (deacquisitionCallback != null) {
+                    if (!deacquisitionCallback.apply(tokenResponse)) {
+                        log.debug("Deacquisition requested abort");
+                        return -1L;
+                    }
+                }
+                // Request a new token, informing the sequencer we were
+                // overwritten.
+                tokenResponse = runtime.getSequencerView()
```

Comment:
I this case, since you obtain a new token, the rank should not be incremented; it is like a normal write, using rank=0.

I think that the wiki is a bit mis-leading regarding the 1-QR "recovery" write. It is really not that different from hole-write. I think we want to leave it out of the protocol, and just implement hole-fill correctly.